### PR TITLE
DRA scheduler: ResourceSlice update

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Making unschedulable pods schedulable again after ResourceSlice cluster events
was accidentally left out when adding structured parameters to Kubernetes 1.30.

#### Special notes for your reviewer:

All E2E tests were defined so that a driver starts first. A new test with a
different order (create pod first, wait for unschedulable, start driver)
triggered the bug and now passes.

While re-reviewing the event handler code, some drive-by changes were also made:
- reduce log verbosity in two cases
- fix feature gate check (irrelevant in practice, see commit message)

#### Does this PR introduce a user-facing change?
```release-note
DRA: when a DRA driver was started after creating pods which need resources from that driver, no additional attempt was made to schedule such unschedulable pods again. Only affected DRA with structured parameters.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 1]
```diff
@@ -491,6 +493,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
 	return framework.Queue, nil
 }
 
+// isSchedulableAfterClaimChange is invoked for add and update claim events reported by
```
Comment (aojea) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:496]: nit s/isSchedulableAfterClaimChange/isSchedulableAfterResourceSliceChange/

**Comment by pohly** [comment]
/retest

pull-kubernetes-e2e-gce failed to come up. Other tests are passing now.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go (hunk 1 above)
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go]: Fixed.

**Comment by aojea** [comment]
not familiar enough with these codebase, it looks correct by diffing against the code around 
LGTM

**Comment by pohly** [comment]
/assign @sanposhiho @alculquicondor 

@aojea did some preliminary review, but someone from SIG Scheduling has to have the final word and needs to approve.

**Comment by alculquicondor** [comment]
/assign @macsko
for first pass

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 2]
```diff
@@ -445,7 +447,7 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
 		// This is not an unexpected error: we know that
 		// foreachPodResourceClaim only returns errors for "not
 		// schedulable".
-		logger.V(4).Info("pod is not schedulable", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
+		logger.V(6).Info("pod is not schedulable", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:450]:
```suggestion
		logger.V(6).Info("Pod is not schedulable after claim change", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
```

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 3]
```diff
@@ -491,6 +493,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
...
+	if err != nil {
+		// Shouldn't happen.
+		return framework.Queue, fmt.Errorf("unexpected object in isSchedulableAfterResourceSliceChange: %w", err)
+	}
+
+	if err := pl.foreachPodResourceClaim(pod, nil); err != nil {
+		// This is not an unexpected error: we know that
+		// foreachPodResourceClaim only returns errors for "not
+		// schedulable".
+		logger.V(6).Info("pod is not schedulable", "pod", klog.KObj(pod), "resourceSlice", klog.KObj(modifiedSlice), "reason", err.Error())
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:511]:
```suggestion
		logger.V(6).Info("Pod is not schedulable after resource slice change", "pod", klog.KObj(pod), "resourceSlice", klog.KObj(modifiedSlice), "reason", err.Error())
```

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 4]
```diff
@@ -491,6 +493,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
 	return framework.Queue, nil
 }
 
+// isSchedulableAfterResourceSliceChange is invoked for add and update slice events reported by
+// an informer. It checks whether that change made a previously unschedulable
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:497]: Explain what exactly could make it schedulable according to this function.

It looks like we just care whether the pod claims exist, but we don't check any correlation between the claims and the ResourceSlice

File: test/e2e/dra/deploy.go [hunk 5]
```diff
@@ -77,69 +77,81 @@ type Nodes struct {
...
-		}()
+		nodes.beforeEach(ctx, f, minNodes, maxNodes)
 	})
 	return nodes
 }
 
+// NewNodesNow is a variant of NewNodes which can be used inside a ginkgo.It.
+func NewNodesNow(ctx context.Context, f *framework.Framework, minNodes, maxNodes int) *Nodes {
+	nodes := &Nodes{}
+	nodes.beforeEach(ctx, f, minNodes, maxNodes)
```
Comment (alculquicondor) [inline test/e2e/dra/deploy.go:93]:
```suggestion
	nodes.newNodes(ctx, f, minNodes, maxNodes)
```
?

Any other name that could summarize the logic in here?

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go (hunk 3 above)
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:511]: Added, and also extended the same message in `isSchedulableAfterClaimChange`.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go (hunk 4 above)
Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:497]: Updated.

We had this discussion on Slack (?) about the tradeoff between execution time for queuing hints vs. accuracy. Based on that discussion, I'm erring on the side of faster execution.

File: test/e2e/dra/deploy.go (hunk 5 above)
Comment (pohly) [reply test/e2e/dra/deploy.go:93]: How about "init"?

**Comment by pohly** [comment]
@alculquicondor: please take another look: https://github.com/kubernetes/kubernetes/compare/fb739de3cb27d5f8f5c21a9273b8be6d23c25bfb..e85d3babf0ff3c010dd8bae9329a2465c2c46527

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Comment by pohly** [comment]
Let's also backport: https://github.com/kubernetes/kubernetes/pull/126873

**Pull request merged**